src/ray/raylet/worker_pool.h

```diff
@@ -684,6 +689,128 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
   /// Agent manager.
   std::shared_ptr<AgentManager> agent_manager_;
 
+  /// Manage all runtime env resources locally by URI reference. We add or remove URI
+  /// reference in the cases below:
```

Comment:
I think this is close! But I'd like to also see sequence of worker states changes. For example, here are possible events for worker lifecycle (not sure it is correct. It is just an example).

```
OnJobStarted  (+1 reference)
OnJobFinished (-1 reference)
OnProcessStarted   (+1 reference)
OnWorkerTimedOut   (-1 reference)
OnWorkerRegistered (-1 reference -> +1 reference)
DisconnectWorker   (-1 reference)
```

And I'd like to see the state change flow. For example, is it possible to have OnProcessStarted -> DisconnectWorker -> OnWorkerTimedOut? If so, we need to handle the edge case (in this case, the ref count will be -1). 

```
# Only one direction is possible
OnJobStarted -> OnJobFinished
```

```
OnProcessStarted -----------------------------
             |                                | -> DisconnectWorker
OnWorkerTimedOut / OnWorkerRegistered --------
```

If this is the possible state flow, we need to unit test

```
OnProcessStarted -> DisconnectWorker
OnProcessStarted -> OnWorkerTimedOut -> DisconnectWorker
OnProcessStarted -> OnWorkerRegistered -> DisconnectWorker
```

So here, the next question is if the # of calls are guaranteed? Fore example, is it possible?
```
OnProcessStarted -> OnWorkerRegistered X 3 -> DisconnectWorker
```

In a glance, this might be possible when there are more than 1 core worker in a process?

Reply:
`OnProcessStarted -> DisconnectWorker` I think we shouldn't add this case because `DisconnectWorker ` doesn't happen when there is no worker registration.

Reply:
`OnProcessStarted -> OnWorkerTimedOut -> DisconnectWorker` I can add this case.

Reply:
@rkooo567 See this test https://github.com/ray-project/ray/pull/20789/commits/c0178b45f2bb8bb7f047b6a092c5ab52b02baabc#diff-bf9523ff768067f9793ae89d37ff19c53c94ff450e39fd20eaf23a57dd38ccbcR1466